refactor(pxe): decompose PXEOracleInterface

mverzilli opened the pull request:
Deprecate `PXEOracleInterface` in favor of more granular and explicit dependencies.

I apologize in advance for the length of this PR, the commits show the progression that led to this final form, and prove that the end result didn't come to me OTOH, but it was the result of dozens of stepwise refactors. At the current state it is possible to spin multiple little PRs off this one if that's deemed saner, it just wasn't clear that would be the case when I originally set out to do this.

## New service layer 

To replace `PXEOracleInterface` (aka `ExecutionDataProvider`), and at the same time keep oracle classes relatively lean and concerns separated, I'm introducing a service layer on top of the existing providers layer.

Due to this PR being big enough as is, and there being more pressing issues to address, I'm not applying this idea to completion, just using it where it helps me untangle things. I do think going further in this direction would help make everything be more unit testable.

The core rules of thumb behind this (long term) are:

- Providers shouldn't talk to each other. This is a simplicity forcing function. Providers are database abstractions.
- Services can talk to each other, and to the provider(s) they encapsulate. Each provider should only be accessed by one service, but one service can manage multiple data providers. So: services get other read/write data needs met by talking to other services.
- The rest of the app talks to services instead of directly to providers. Again, I'm not tackling that here because the PR is big enough as is, but it's an architecture we could slowly move towards.

## Capsules as plumbing

Something that became clear along the way of this refactor is that capsule state management in many cases is a transport concern, and as such code becomes much more cohesive if we separate the code that deals with that, from the code that actually operates on the contents. Maybe the nicest example of that can be seen in `utilityBulkRetrieveLogs`.

## Working with Anchor blocks

I think most of the places where we now consume `AnchorBlockDataProvider` to get the current anchor block, should be modified to instead receive the anchor block as param, which will make it easier to see that in reality the anchor block is pinned for a given `ContractFunctionSimulator` activation. That, in turn, should make the work on DB integrity easier, and it will also be helpful to eventually adopt a less conservative approach towards concurrency. 

## Temporary complexity increase (waiting for new tagging sync work)

There are some parts of the app that might look a bit more convoluted with this refactor, mostly due to porting sync tagging functionality to make things compile and pass tests. Note that @benesjan is overhauling sync tagging, so I'm deliberately not investing effort on simplifying those parts and will instead wait and see what's necessary once that work is available.  

Some specific cases of this situation are found on:
- Test setup of `oracle_version_is_checked.test.ts`: it got much more complex given `ContractFunctionSimulator` now receives all the data providers it needs to operate instead of hiding everything under `PXEOracleInterface`. A lot of this complexity is probably derived from the old tag sync'ing logic and I'm expecting it to just disintegrate.
- `private_execution.test.ts`: I'm hoping a lot of the boilerplate will be unnecessary once the new work on tagging lands.
- `LogService` and its tests

Fixes #17776

mverzilli commented on playground/src/utils/artifacts.ts:
```diff
@@ -0,0 +1,27 @@
+    const decompressor = new DecompressionStream('gzip');
+    const sourceStream = response.body ?? new Blob([await response.arrayBuffer()]).stream();
+    return new Response(sourceStream.pipeThrough(decompressor)).text();
+  }
```
`DecompressionStream` is not available on Safari so this works it around

mverzilli commented on playground/package.json:
```diff
@@ -34,6 +34,7 @@
     "@uidotdev/usehooks": "^2.4.1",
     "buffer-json": "^2.0.0",
     "nosleep.js": "^0.12.0",
+    "pako": "^2.1.0",
```
to deal with client side decompression

mverzilli commented on playground/vite.config.ts:
```diff
@@ -101,6 +101,61 @@ export default defineConfig(({ mode }) => {
+    build: {
+      rollupOptions: {
+        output: {
+          manualChunks(id) {
```
absolute file path of module

mverzilli commented on playground/vite.config.ts:
```diff
@@ -101,6 +101,61 @@ export default defineConfig(({ mode }) => {
+      rollupOptions: {
+        output: {
+          manualChunks(id) {
+            const normalizedId = id.split('?')[0];
```
removes query strings

mverzilli commented on playground/vite.config.ts:
```diff
@@ -101,6 +101,61 @@ export default defineConfig(({ mode }) => {
+                return `aztec-${packageName}-artifact-${path.basename(normalizedId, '.json')}`;
+              }
+              return undefined;
+            };
```
produces very fine-grained chunks like:

aztec-foo-artifact-BarContract

mverzilli commented on playground/vite.config.ts:
```diff
@@ -101,6 +101,61 @@ export default defineConfig(({ mode }) => {
+                if (artifactChunk) return artifactChunk;
+                return `aztec-${packageName}`;
+              }
+            }
```
creates one chunk per yarn-project/package-name

benesjan commented on yarn-project/pxe/src/contract_function_simulator/oracle/private_execution.ts:
```diff
@@ -187,14 +187,19 @@ export async function readCurrentClassId(
  */
 export async function verifyCurrentClassId(
```
The docs above our outdated now.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/oracle/private_execution_oracle.ts:
```diff
@@ -79,15 +89,41 @@ export class PrivateExecutionOracle extends UtilityExecutionOracle implements IP
     private readonly noteCache: ExecutionNoteCache,
     private readonly taggingIndexCache: ExecutionTaggingIndexCache,
-    executionDataProvider: ExecutionDataProvider,
+    contractDataProvider: ContractDataProvider,
+    noteDataProvider: NoteDataProvider,
+    keyStore: KeyStore,
+    addressDataProvider: AddressDataProvider,
+    aztecNode: AztecNode,
+    anchorBlockDataProvider: AnchorBlockDataProvider,
+    senderTaggingDataProvider: SenderTaggingDataProvider,
+    recipientTaggingDataProvider: RecipientTaggingDataProvider,
+    capsuleDataProvider: CapsuleDataProvider,
+    privateEventDataProvider: PrivateEventDataProvider,
```
Nico originally had an idea of wrapping data providers in some kind of Providers class that could be passed around. It would make the code here cleaner.

WDYT?

mverzilli replied:
I'm a bit on the fence because it can obscure things (kind of how PXEOracleInterface ended up seeming a good place to put things because there were things in it).

I think I'd rather eventually explore using TS decorators, which we could use to say "this constructor expects these providers"  (https://www.typescriptlang.org/docs/handbook/decorators.html)

mverzilli replied:
In any case, it's a discussion we can defer for now

benesjan commented on yarn-project/pxe/src/contract_function_simulator/oracle/utility_execution_oracle.ts:
```diff
@@ -32,13 +54,24 @@ export class UtilityExecutionOracle implements IMiscOracle, IUtilityExecutionOra
     protected readonly capsules: Capsule[], // TODO(#12425): Rename to transientCapsules
     protected readonly anchorBlockHeader: BlockHeader,
-    protected readonly executionDataProvider: ExecutionDataProvider,
+    protected readonly contractDataProvider: ContractDataProvider,
+    protected readonly noteDataProvider: NoteDataProvider,
+    protected readonly keyStore: KeyStore,
+    protected readonly addressDataProvider: AddressDataProvider,
+    protected readonly aztecNode: AztecNode,
+    protected readonly anchorBlockDataProvider: AnchorBlockDataProvider,
+    protected readonly senderTaggingDataProvider: SenderTaggingDataProvider,
+    protected readonly recipientTaggingDataProvider: RecipientTaggingDataProvider,
+    protected readonly capsuleDataProvider: CapsuleDataProvider,
+    protected readonly privateEventDataProvider: PrivateEventDataProvider,
```
This seems to be all over the places. I think a wrapper class would make it all nicer.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/oracle/oracle_version_is_checked.test.ts:
```diff
@@ -1,47 +1,119 @@
 
   beforeEach(async () => {
-    executionDataProvider = mock<ExecutionDataProvider>();
+    contractDataProvider = mock<ContractDataProvider>();
+    noteDataProvider = mock<NoteDataProvider>();
+    keyStore = mock<KeyStore>();
+    addressDataProvider = mock<AddressDataProvider>();
+    aztecNode = mock<AztecNode>();
+    anchorBlockDataProvider = mock<AnchorBlockDataProvider>();
+    senderTaggingDataProvider = mock<SenderTaggingDataProvider>();
+    recipientTaggingDataProvider = mock<RecipientTaggingDataProvider>();
+    capsuleDataProvider = mock<CapsuleDataProvider>();
+    privateEventDataProvider = mock<PrivateEventDataProvider>();
+    utilityAssertCompatibleOracleVersionSpy = jest.spyOn(
+      UtilityExecutionOracle.prototype,
+      'utilityAssertCompatibleOracleVersion',
+    );
+    utilityAssertCompatibleOracleVersionSpy.mockClear();
```
If we had `Providers` wrapper class maybe it could then make these mocks easier - maybe we could then have a helper that would mock it the providers and set mock functions on the providers that return "null" values.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/proxied_contract_data_source.ts:
```diff
@@ -52,6 +52,23 @@ export class ProxiedContractDataProviderFactory {
               }
             };
           }
+          case 'getFunctionArtifactWithDebugMetadata': {
```
Why has this appeared here?

mverzilli replied:
`getFunctionArtifact` used to be in `PXEOracleInterface`, which called `getFunctionArtifact` on `ContractDataProvider` and augmented it with debug data. 

So I moved that logic to a new function in `ContractDataProvider` called `getFunctionArtifactWithDebugMetadata` (which is consumed from `ContractFunctionSimulator`).

What I'm not 100% sure is that the overriding ofr `getFunctionArtifact` a few lines above is still needed

benesjan replied:
> What I'm not 100% sure is that the overriding ofr getFunctionArtifact a few lines above is still needed

Think it's ok to just leave it there.

benesjan commented on yarn-project/pxe/src/logs/log_service.ts:
```diff
@@ -0,0 +1,387 @@
+import { Tag } from '../tagging/tag.js';
+import { getInitialIndexesMap, getPreTagsForTheWindow } from '../tagging/utils.js';
+
+// TODO: wanted: good name
```
Why do you think the name is bad?

mverzilli replied:
`LogService` is two very overloaded words together, but yeah, it's not terrible

benesjan commented on yarn-project/pxe/src/tree_membership/tree_membership_service.ts:
```diff
@@ -0,0 +1,112 @@
+
+import type { AnchorBlockDataProvider } from '../storage/anchor_block_data_provider/anchor_block_data_provider.js';
+
+export class TreeMembershipService {
```
Unrelated to this PR but this service is so light that it feels like this should be provided just by `AztecNode`.

In the future would consider whether we can adjust AztecNode API and drop this.

Would you agree?

mverzilli replied:
Absolutely. The only reason for this to exist is to augment AztecNode without changing its API, which would turn this PR into something even bigger than it is already. The moment we don't need this code the service loses its right to exist :)

benesjan reviewed (approved):
LGTM

As discussed would introduce the `Providers` type in a followup PR as that would allow us to pass all the providers around and possibly simplify the mocks in tests.

The services approach seems to be a great idea!

mverzilli committed:
refactor(pxe): decompose PXEOracleInterface

Deprecate `PXEOracleInterface` in favor of more granular and explicit dependencies.

I apologize in advance for the length of this PR, the commits show the progression that led to this final form, and prove that the end result didn't come to me OTOH, but it was the result of dozens of stepwise refactors. At the current state it is possible to spin multiple little PRs off this one if that's deemed saner, it just wasn't clear that would be the case when I originally set out to do this.

## New service layer

To replace `PXEOracleInterface` (aka `ExecutionDataProvider`), and at the same time keep oracle classes relatively lean and concerns separated, I'm introducing a service layer on top of the existing providers layer.

Due to this PR being big enough as is, and there being more pressing issues to address, I'm not applying this idea to completion, just using it where it helps me untangle things. I do think going further in this direction would help make everything be more unit testable.

The core rules of thumb behind this (long term) are:

- Providers shouldn't talk to each other. This is a simplicity forcing function. Providers are database abstractions.
- Services can talk to each other, and to the provider(s) they encapsulate. Each provider should only be accessed by one service, but one service can manage multiple data providers. So: services get other read/write data needs met by talking to other services.
- The rest of the app talks to services instead of directly to providers. Again, I'm not tackling that here because the PR is big enough as is, but it's an architecture we could slowly move towards.

## Capsules as plumbing

Something that became clear along the way of this refactor is that capsule state management in many cases is a transport concern, and as such code becomes much more cohesive if we separate the code that deals with that, from the code that actually operates on the contents. Maybe the nicest example of that can be seen in `utilityBulkRetrieveLogs`.

## Working with Anchor blocks

I think most of the places where we now consume `AnchorBlockDataProvider` to get the current anchor block, should be modified to instead receive the anchor block as param, which will make it easier to see that in reality the anchor block is pinned for a given `ContractFunctionSimulator` activation. That, in turn, should make the work on DB integrity easier, and it will also be helpful to eventually adopt a less conservative approach towards concurrency.

## Temporary complexity increase (waiting for new tagging sync work)

There are some parts of the app that might look a bit more convoluted with this refactor, mostly due to porting sync tagging functionality to make things compile and pass tests. Note that @benesjan is overhauling sync tagging, so I'm deliberately not investing effort on simplifying those parts and will instead wait and see what's necessary once that work is available.

Some specific cases of this situation are found on:
- Test setup of `oracle_version_is_checked.test.ts`: it got much more complex given `ContractFunctionSimulator` now receives all the data providers it needs to operate instead of hiding everything under `PXEOracleInterface`. A lot of this complexity is probably derived from the old tag sync'ing logic and I'm expecting it to just disintegrate.
- `private_execution.test.ts`: I'm hoping a lot of the boilerplate will be unnecessary once the new work on tagging lands.
- `LogService` and its tests

Fixes #17776
